libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -876,6 +952,125 @@ static IotMqttError_t _waitForOperation( IotMqttOperation_t operation,
 
 /*-----------------------------------------------------------*/
 
+static IotMqttError_t _scheduleKeepAlive( IotMqttConnection_t pMqttConnection )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+    IotTaskPoolError_t taskPoolStatus = IOT_TASKPOOL_SUCCESS;
+
+    /* Check if a keep-alive job should be scheduled. */
+    if( pMqttConnection->pingreq.u.operation.periodic.ping.keepAliveMs != 0U )
+    {
+        IotLogDebug( "Scheduling first MQTT keep-alive job." );
+
+        taskPoolStatus = IotTaskPool_ScheduleDeferred( IOT_SYSTEM_TASKPOOL,
+                                                       pMqttConnection->pingreq.job,
+                                                       pMqttConnection->pingreq.u.operation.periodic.ping.nextPeriodMs );
+    }
+
+    if( taskPoolStatus != IOT_TASKPOOL_SUCCESS )
+    {
+        status = IOT_MQTT_SCHEDULING_ERROR;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotMqttError_t _sendConnectRequest( _mqttOperation_t * pOperation,
+                                           uint32_t timeoutMs )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+
+    IotMqtt_Assert( pOperation != NULL );
+
+    /* Check the serialized MQTT packet. */
+    IotMqtt_Assert( pOperation->u.operation.pMqttPacket != NULL );
+    IotMqtt_Assert( pOperation->u.operation.packetSize > 0U );
+
+    /* Send the CONNECT packet. */
+    _IotMqtt_ProcessSend( IOT_SYSTEM_TASKPOOL, pOperation->job, pOperation );
+
+    /* Wait for the CONNECT operation to complete, i.e. wait for CONNACK. */
+    status = IotMqtt_Wait( pOperation, timeoutMs );
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotMqttError_t _addSubscriptions( _mqttOperation_t * pOperation,
+                                         IotMqttConnection_t pMqttConnection,
+                                         const IotMqttConnectInfo_t * pConnectInfo )
+{
+    IotMqttError_t status = IOT_MQTT_SUCCESS;
+
+    IotMqtt_Assert( pOperation != NULL );
+    IotMqtt_Assert( pConnectInfo != NULL );
+
+    /* Ensure the members set by operation creation and serialization
+     * are appropriate for a blocking CONNECT. */
+    IotMqtt_Assert( pOperation->u.operation.status == IOT_MQTT_STATUS_PENDING );
+    IotMqtt_Assert( ( pOperation->u.operation.flags & IOT_MQTT_FLAG_WAITABLE )
+                    == IOT_MQTT_FLAG_WAITABLE );
+    IotMqtt_Assert( pOperation->u.operation.periodic.retry.limit == 0U );
+
+    /* Set the operation type. */
+    pOperation->u.operation.type = IOT_MQTT_CONNECT;
+
+    /* Add previous session subscriptions. */
+    if( pConnectInfo->pPreviousSubscriptions != NULL )
+    {
+        /* Previous subscription count should have been validated as nonzero. */
+        IotMqtt_Assert( pConnectInfo->previousSubscriptionCount > 0U );
+
+        status = _IotMqtt_AddSubscriptions( pMqttConnection,
+                                            2,
```

Comment:
This should be a `#define` or `enum` rather than a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)), especially since this is part of the MQTT spec.  It looks like the closest match in the current code is [`MQTT_PACKET_TYPE_SUBSCRIBE`](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta/libraries/standard/mqtt/src/private/iot_mqtt_internal.h#L256), however this is `0x82` (that's also [the value in the spec](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718063))rather than 2 as specified here.  Please add a `#define` or `enum` for this value.

Reply:
This is not  setting 'packet type' , but setting packet identifier. It can have any unique value to identify a packet.  This function is adding existing subscriptions to the managed subscription internal list, therefore therefore it can be any value.